pgjdbc/src/main/java/org/postgresql/util/LazyCleaner.java

```diff
@@ -0,0 +1,196 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.postgresql.util;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class LazyCleaner {
+  private static final Logger LOGGER = Logger.getLogger(LazyCleaner.class.getName());
+
+  public interface Cleanable {
+    void clean();
+  }
+
+  public interface CleaningAction<T extends Throwable> {
+    void onClean(boolean leak) throws T;
+  }
+
+  private final ReferenceQueue<Object> queue = new ReferenceQueue<Object>();
+  private final long threadTtl;
+  private final ThreadFactory threadFactory;
+  private boolean threadRunning = false;
+  private int watchedCount = 0;
+  private @Nullable Node first;
+  private @Nullable Cleanable keepAliveCleanable;
+
+  public LazyCleaner(long threadTtl, final String threadName) {
+    this(threadTtl, new ThreadFactory() {
+      public Thread newThread(Runnable runnable) {
+        Thread thread = new Thread(runnable, threadName);
+        thread.setDaemon(true);
+        return thread;
+      }
+    });
+  }
+
+  public LazyCleaner(long threadTtl, ThreadFactory threadFactory) {
+    this.threadTtl = threadTtl;
+    this.threadFactory = threadFactory;
+  }
+
+  public synchronized LazyCleaner setKeepThreadAlive(boolean alive) {
+    if (alive == (keepAliveCleanable != null)) {
+      return this;
+    }
+
+    if (alive) {
+      keepAliveCleanable = register(this, null);
+    } else {
+      if (keepAliveCleanable != null) {
+        keepAliveCleanable.clean();
+      }
+      keepAliveCleanable = null;
+    }
+    return this;
+  }
+
+  public Cleanable register(Object obj, @Nullable CleaningAction action) {
+    return add(new Node(obj, action));
+  }
+
+  public int getWatchedCount() {
+    return watchedCount;
+  }
+
+  public boolean isThreadRunning() {
+    return threadRunning;
+  }
+
+  private synchronized boolean checkEmpty() {
+    if (first == null) {
+      threadRunning = false;
+      return true;
+    }
+    return false;
+  }
+
+  private synchronized Node add(Node node) {
+    if (first != null) {
+      node.next = first;
+      first.prev = node;
+    }
+    first = node;
+    watchedCount++;
+
+    if (!threadRunning) {
+      threadRunning = startThread();
+    }
+    return node;
+  }
+
+  private boolean startThread() {
+    Thread thread = threadFactory.newThread(new Runnable() {
```

Comment:
Doesn't this create a lot of threads? How about using an ExecutorService to handle the runnables with a fixed number of threads?

Reply:
Would you suggest an alternative that supports class unloading?

Reply:
By the way, the thread exits only in case all the connections are closed/abandoned, so it should not  impact applications

Reply:
Unloading which classes?

Reply:
org.postgresql

Reply:
I'm sorry. I can't follow why you need to unload classes for org.postgresql.

Reply:
org.postgresql classes should support unloading to prevent memory leaks

Reply:
Here's a sample issue: https://github.com/pgjdbc/pgjdbc/pull/188

Reply:
Ok. I think I see. The executor threads wouldn't reference any postgres classes. The queued runnables would. If those runnables get removed from the queue there would be no leak.